pkg/provisioner/ironic/ironic.go

```diff
@@ -546,6 +546,36 @@ func (p *ironicProvisioner) ValidateManagementAccess(credentialsChanged, force b
 	// 	return result, errors.Wrap(err, "failed to get provisioning state in ironic")
 	// }
 
+	p.log.Info("setting automated cleaning mode to",
+		"ID", ironicNode.UUID,
+		"mode", p.host.Spec.AutomatedCleaningMode)
+
+	if (p.host.Spec.AutomatedCleaningMode == "disabled" && *ironicNode.AutomatedClean != false) || (p.host.Spec.AutomatedCleaningMode == "enabled" && *ironicNode.AutomatedClean == false) {
+		value := p.host.Spec.AutomatedCleaningMode != metal3v1alpha1.CleaningModeDisabled
+
+		updates := nodes.UpdateOpts{
```

Comment:
Does this need to be an additional update?

I was thinking this could be called immediately after `getImageUpdateOptsForNode` (e.g a new `getCleaningModeOpts ` function perhaps?) and appended to the `updates`?

@zaneb am I right in saying we should hit that code in both the Provisioning and Deprovisioning state, since imageData will be set via either the Spec or Status?

Reply:
I [agree](https://github.com/metal3-io/baremetal-operator/pull/784#discussion_r597888586), we should stop copying and pasting the code to Update() and all of its error handling, which we already call in 3 places (now 4) when 1 would suffice, and just add one more thing to the list.

Note, however, that the `getImageUpdateOptsForNode()` is only called if the node doesn't exist. Which is a bug, because if updating it fails then it will never get called again. You are correct that the image data is available in both the Provisioning and Deprovisioning states.

We could refactor and fix the bug in a separate PR if that helps.

Reply:
Do we need a new func for this? or can we just add it within  `getImageUpdateOptsForNode()?`

Reply:
I would just do the Update() once after the if/else, and have the stuff that is currently calling Update() just add stuff to the updates list.
We should definitely not call Update() within `getImageUpdateOptsForNode()` with the same list of updates that we're about to return, only for Update() to be called with it again.

Reply:
@zaneb can you please take a look if the current state is what you meant?

Reply:
I wouldn't put it inside `getImageUpdateOptsForNode()` because:

1. of the bug mentioned above, which means it is not called most of the time when we want it to be called (you'd have to delete the ironic DB to get the cleaning mode to update)
2. even when the bug is fixed, it'll only be called when there is an image specified, but that's not a condition we want to impose on whether the cleaning mode gets updated